### PR TITLE
Merge demo branch to master

### PR DESCRIPTION
**Pull request by jensj12**
Don't do anything else until this is finished.

### CODE REVIEW

File: MeesGame/Gameplay/Editor/MazeGenerator.cs [hunk 1]
```diff
@@ -102,10 +106,16 @@ public static MeesGame.TileField GenerateMaze(MeesGame.TileField tiles, int star
                     nodesToDo.RemoveAt(position);
                 }
             }
-
+            tiles.Add(new MeesGame.EndTile(), 0, 0);
             return tiles;
         }
 
+        public static TileField GenerateMazeWithRandomStart()
```
Comment (jensj12) [inline MeesGame/Gameplay/Editor/MazeGenerator.cs]: The generator should always do this. The level should determine the start from the tileField data.

File: MeesGame/Gameplay/GameManagement/GameObject.cs [hunk 2]
```diff
@@ -3,11 +3,11 @@
...
 public abstract class GameObject : IGameLoopObject
 {
-    protected GameObject parent;
-    protected Vector2 position, velocity;
-    protected int layer;
-    protected string id;
-    protected bool visible;
+    public GameObject parent;
+    public Vector2 position, velocity;
+    public int layer;
```
Comment (jensj12) [inline MeesGame/Gameplay/GameManagement/GameObject.cs]: It's no longer needed to make these public IIRC

File: MeesGame/Gameplay/GameManagement/InputHelper.cs [hunk 3]
```diff
@@ -6,11 +6,13 @@ public class InputHelper
     protected MouseState currentMouseState, previousMouseState;
     protected KeyboardState currentKeyboardState, previousKeyboardState;
     protected Vector2 scale, offset;
+    protected int scrollOld, scrollNew;
```
Comment (jensj12) [inline MeesGame/Gameplay/GameManagement/InputHelper.cs]: Please keep the naming consistent with the other variables.

File: MeesGame/Gameplay/GameObjects/Tile.cs [hunk 4]
```diff
@@ -223,6 +285,20 @@ public virtual InventoryItem GetItem()
         {
             return null;
         }
+
+        public abstract void UpdateGraphics();
```
Comment (jensj12) [inline MeesGame/Gameplay/GameObjects/Tile.cs:301]: Name should be more descriptive and it's missing a summary.

File: MeesGame/Gameplay/GameObjects/HoleTile.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        new public static string[] GetDefaultAssetNames()
+        {
+            string[] parentDefaultAssetNames = FloorTile.GetDefaultAssetNames();
+            string[] defaultAssetNames = new string[parentDefaultAssetNames.Length + 1];
+            parentDefaultAssetNames.CopyTo(defaultAssetNames, 0);
+            defaultAssetNames[parentDefaultAssetNames.Length] = defaultAssetName;
+            return defaultAssetNames;
+        }
+
+        public bool FallInHole(Player player)
```
Comment (jensj12) [inline MeesGame/Gameplay/GameObjects/HoleTile.cs]: Should use EnterTile function (or whatever is defined in the abstract Tile class)

File: MeesGame/Gameplay/GameObjects/TileField.cs [hunk 6]
```diff
@@ -90,7 +108,15 @@ public void UpdateGraphicsToMatchSurroundings()
             }
         }
 
-        public void revealArea(Point a)
+        public void UpdateGraphics()
```
Comment (jensj12) [inline MeesGame/Gameplay/GameObjects/TileField.cs:111]: Name should be more descriptive

File: MeesGame/Gameplay/GameStates/PlayingLevelState.cs [hunk 7]
```diff
@@ -36,38 +35,50 @@ private void initOverlay()
...
         }
 
         public void Reset()
         {
-            level[currentLevelIndex].Reset();
             overlay.Reset();
+            InitOverlay();
         }
     }
 }
```
Comment (jensj12) [inline MeesGame/Gameplay/GameStates/PlayingLevelState.cs]: Newline at end is missing.

File: MeesGame/Gameplay/Level/EditorLevel.cs [hunk 8]
```diff
@@ -10,14 +10,22 @@ public EditorLevel(int levelindex = 0, int screenWidth = -1, int screenHeight =
...
+        }
+
+        public void FillLevelWithTiles(TileField tiles)
+        {
+            for (int x = 0; x < numRows; x++)
+                for (int y = 0; y < numColumns; y++)
+                    this.Tiles.Add(tiles.Objects[x,y], x, y);
         }
     }
 }
```
Comment (jensj12) [inline MeesGame/Gameplay/Level/EditorLevel.cs]: Missing newline

File: MeesGame/Gameplay/UIObjects/Objects/FileExplorer.cs [hunk 9]
```diff
@@ -75,9 +78,21 @@ public void generateFileList()
...
+        public string SelectedFile
+        {
+            get
+            {
+                if (selected == NOTHING_SELECTED) return null;
+                return fileList[selected];
+            }
         }
     }
 }
```
Comment (jensj12) [inline MeesGame/Gameplay/UIObjects/Objects/FileExplorer.cs]: Missing newline

File: MeesGame/Gameplay/GameObjects/Tile.cs (hunk 4 above)
Comment (maxmeijer42) [reply MeesGame/Gameplay/GameObjects/Tile.cs:301]: Is this function really necessary? It's currently only used in KeyTile but KeyTile could just update the graphics in the EnterTile() function

File: MeesGame/Gameplay/GameStates/GameOverState.cs [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        //No need to update anything, it's game over
+        public void Update(GameTime gameTime)
+        {
+            overlay.Update(gameTime);
+            return;
+        }
+
+        //Draw the correct gameover overlay
+        //TODO: Add a draw function once the overlay is here
+        public void Draw(GameTime gameTime, SpriteBatch spriteBatch)
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/GameStates/GameOverState.cs:34]: obsolete TODOs

File: MeesGame/Gameplay/Level/FileIO.cs [hunk 11]
```diff
@@ -10,29 +11,37 @@ class FileIO
         /// <summary>
         /// Saves the level to a file with the specified name.
         /// </summary>
-        /// <param name="obj"> Dummy variable. </param>
-        /// <param name="ea"> Dummy variable. </param>
-        public void Save(object obj, EventArgs ea)
+        /// <param name="level"> The level being saved. </param>
+        /// <param name="filename"> The name of the file the level is being saved to. </param>
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/Level/FileIO.cs]: obsolete doc

File: MeesGame/Gameplay/GameObjects/Player.cs [hunk 12]
```diff
@@ -132,6 +132,12 @@ public void Win()
             PlayerWin?.Invoke(this);
         }
 
+        // TODO: make the player lose instead of win
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/GameObjects/Player.cs:135]: 👍

**Review by maxmeijer42 — APPROVED** [review]

**Pull request merged**